modules/audio/src/main/java/com/gluonhq/attach/audio/impl/AndroidAudioService.java

```diff
@@ -0,0 +1,236 @@
+     * @return file name in private storage if file was copied or if it already exists at destination,
+     * or empty String if there are any errors during copying
+     */
+    private String prepareFile(URL url, String subDirName) {
+        if (privateStorage == null) {
+            privateStorage = setUpDirectories();
+        }
+
+        String extForm = url.toExternalForm();
+        String fileName = extForm.substring(extForm.lastIndexOf("/") + 1);
+        String fullName = privateStorage.getAbsolutePath() + "/assets/" + subDirName + fileName;
+
+        File outputFile = new File(fullName);
+
+        if (!outputFile.exists()) {
+            try {
+                copyFile(url, outputFile);
+            } catch (Exception e) {
+
+                // TODO:
+                e.printStackTrace();
+
+                return "";
+            }
+        }
+
+        return fullName;
+    }
+
+    private File setUpDirectories() {
+        File storage = Services.get(StorageService.class)
+                .flatMap(service -> service.getPrivateStorage())
+                .orElseThrow(() -> new RuntimeException("Error accessing Private Storage folder"));
+
+        File assetsDir = new File(storage, "assets");
+        if (!assetsDir.exists()) {
+            assetsDir.mkdir();
+        }
+
+        File musicDir = new File(assetsDir, "music");
+        if (!musicDir.exists()) {
+            musicDir.mkdir();
+        }
+
+        File soundsDir = new File(assetsDir, "sounds");
+        if (!soundsDir.exists()) {
+            soundsDir.mkdir();
+        }
+
+        return storage;
+    }
+
+    private void copyFile(URL url, File outputFile) throws Exception {
+        try (InputStream input = url.openStream()) {
+            if (input == null) {
+                throw new RuntimeException("Internal copy failed: input stream for " + url + " is null");
+            }
+
+            try (OutputStream output = new FileOutputStream(outputFile)) {
+                byte[] buffer = new byte[1024];
+                int length;
+                while ((length = input.read(buffer)) > 0) {
+                    output.write(buffer, 0, length);
+                }
+                output.flush();
+            }
+        }
+    }
+
+    private static class AndroidAudio implements Audio {
+
+        private boolean isDisposed = false;
+        private final int id;
+
+        AndroidAudio(int id) {
+            this.id = id;
+        }
+
+        @Override
+        public void setLooping(boolean looping) {
+            if (isDisposed)
+                return;
+
+            AndroidAudioService.setLooping(id, looping);
+        }
+
+        @Override
+        public void setVolume(double volume) {
+            if (isDisposed)
+                return;
+
+            AndroidAudioService.setVolume(id, volume);
+        }
+
+        @Override
+        public void setOnFinished(Runnable action) {
+            // TODO:
```

Comment:
For this part you'll need to add a callback from DalvikAudioService (via JNI), so you'll get notified when the audio finishes.

Reply:
Yes, but I'm currently in two minds about whether we actually need `setOnFinished()` considering we don't have a `DalvikSound` implementation for it, just Music... I think it's the same for Desktop? I don't remember if `SoundPool` allows to listen for end of play... Any thoughts?

Reply:
Actually I'm more in favor of adding the Media.Status property, through the required callback, so the user can react to any status change as needed,  and removing `setOnFinished` from the API

Reply:
This makes sense for Music, but I don't know if Sound can provide this info. [SoundPool](https://developer.android.com/reference/android/media/SoundPool) does not seem to provide fine-grained information about audio state.

Reply:
Ok, that seems the case. But you know about play, pause, resume, stop, so those actions can define the media status for Music and for Sound too. 
The completionListener on Video is used only for moving to the next item, which is not required here.

Reply:
Agree on removing `setOnFinished()`.

As for media status, I think we still come back to the same issue of not having enough info. For example, when looping is false, PLAY will eventually _internally_ change to STOP. Since there are no listeners, the state, from our perspective, is still PLAY and not STOP. We can't readily monitor this change with Sound. Perhaps, instead we should think about possible uses cases and then go from there?